### PR TITLE
Allow to apply code fix in top-level statements

### PR DESCRIPTION
**Pull request by meziantou**
When you apply the code fix in the following code, you get an exception:

````c#
using System;

Console.WriteLine("Hello, World");
````

````
System.InvalidOperationException : Sequence contains no elements
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Spectre.Console.Analyzer.CodeActions.SwitchToAnsiConsoleAction.GetAnsiConsoleParameterDeclaration()
   at async Spectre.Console.Analyzer.CodeActions.SwitchToAnsiConsoleAction.GetChangedDocumentAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedSolutionAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.ComputeOperationsAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetPreviewOperationsAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.GetPreviewResultAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionWithNestedFlavors.PreviewChangesSuggestedAction.CreateAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionWithNestedFlavors.GetPreviewChangesFlavorAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionWithNestedFlavors.CreateAllFlavorsAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunctionAsync[T](<Unknown Parameters>)
````

### CODE REVIEW

File: src/Spectre.Console.Analyzer/Fixes/FixProviders/StaticAnsiConsoleToInstanceFix.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
         var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
         if (root != null)
         {
-            var methodDeclaration = root.FindNode(context.Span).FirstAncestorOrSelf<InvocationExpressionSyntax>();
+            var methodDeclaration = root.FindNode(context.Span, getInnermostNodeForTie: true).FirstAncestorOrSelf<InvocationExpressionSyntax>();
```
Comment (meziantou) [inline src/Spectre.Console.Analyzer/Fixes/FixProviders/StaticAnsiConsoleToInstanceFix.cs:23]: I'm not sure this is needed, but as it uses `FirstAncestorOrSelf`, it should find the right statement anyway. Let me know if you want me to revert this change.

File: test/Spectre.Console.Analyzer.Tests/SpectreAnalyzerVerifier.cs [hunk 2]
```diff
@@ -37,7 +37,8 @@ private class Test : CSharpCodeFixTest<TAnalyzer, EmptyCodeFixProvider, XUnitVer
         public Test()
         {
             ReferenceAssemblies = CodeAnalyzerHelper.CurrentSpectre;
-            TestBehaviors |= TestBehaviors.SkipGeneratedCodeCheck;
+            TestBehaviors |= TestBehaviors.SkipGeneratedCodeCheck;
+            CompilerDiagnostics = CompilerDiagnostics.None;
```
Comment (meziantou) [inline test/Spectre.Console.Analyzer.Tests/SpectreAnalyzerVerifier.cs:41]: The new test use top-level statements. This is only valid with `OutputKind.ConsoleApplication`. So, the compiler reports an error (CS8805). I tried to use the following code to remove the diagnostics:

````
        protected override CompilationOptions CreateCompilationOptions()
            => new CSharpCompilationOptions(OutputKind.ConsoleApplication, allowUnsafe: true);
````

But it doesn't seem to be enough. Let me know if there is a better solution.

File: test/Spectre.Console.Analyzer.Tests/Unit/Fixes/UseSpectreInsteadOfSystemConsoleFixTests.cs [hunk 3]
```diff
@@ -1,142 +1,163 @@
...
-        _ansiConsole.WriteLine(""Hello, World"");
-    }
-}";
-
-        await SpectreAnalyzerVerifier<UseSpectreInsteadOfSystemConsoleAnalyzer>
-            .VerifyCodeFixAsync(Source, _expectedDiagnostic.WithLocation(11, 9), FixedSource)
-            .ConfigureAwait(false);
-    }
-}
+namespace Spectre.Console.Analyzer.Tests.Unit.Fixes;
```
Comment (meziantou) [inline test/Spectre.Console.Analyzer.Tests/Unit/Fixes/UseSpectreInsteadOfSystemConsoleFixTests.cs:1]: I think I mess up with CRLF and LF.

There is no `editorconfig` to specify what is desired for EOL. On my machine (Windows) all the tests were failing after I check out the repo. So, I changed the EOL to CRLF using VS to make the tests pass. I think it works on other machines as dev keeps the default option to adjust the EOL automatically. On my machine, I use `core.autocrlf=false`.

Comment (meziantou) [reply test/Spectre.Console.Analyzer.Tests/Unit/Fixes/UseSpectreInsteadOfSystemConsoleFixTests.cs]: That's it. I changed the EOL to LF and now the CI pass.
I'll update my local git repo config to change EOL automatically.

File: test/Spectre.Console.Analyzer.Tests/SpectreAnalyzerVerifier.cs (hunk 2 above)
Comment (meziantou) [reply test/Spectre.Console.Analyzer.Tests/SpectreAnalyzerVerifier.cs]: I found how to set the `OutputKind`. It's in `Test.TestState.OutputKind`

**Comment by phil-scott-78** [comment]
Closed this, merged as part of #1169
